### PR TITLE
Update to latest CUDA, use NCCL_jll

### PR DESCRIPTION
**Pull request by simonbyrne**
This updates NCCL.jl to CUDA.jl 5, and to use the new NCCL_jll build.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Okay, this is now working locally for me: I've tried to keep the argument names/keywords args somewhat consistent with MPI.jl.

Remaining items
- [ ] Set up CI: if you can set up a pipeline, I'm happy to try to figure it out. Note that NCCL appears to require at least 3 devices, see https://github.com/NVIDIA/nccl/issues/1125.
- [ ] Set up Docs

File: src/communicator.jl [hunk 1]
```diff
@@ -1,72 +1,143 @@
...
+communicator `comm`.
+"""
+function default_device_stream(devid::Integer)
+    state = CUDA.task_local_state!()
+    devidx = devid+1
+    @inbounds if state.streams[devidx] === nothing
+        state.streams[devidx] = CUDA.device!(CUDA.create_stream, devid)
+    else
+        state.streams[devidx]::CuStream
+    end
```
Comment (simonbyrne) [inline src/communicator.jl:140]: This depends on the CUDA internals, so might be something worth adding there?

**Comment by simonbyrne** [comment]
@maleadt ?

**Comment by maleadt** [comment]
We don't have enough GPUs to take 3 devices and assign them to a single worker, only for testing of NCCL.jl...

**Comment by simonbyrne** [comment]
Fair enough. Okay, I think this is ready. It may need a `DOCUMENTER_KEY` if it doesn't already exist.

File: src/communicator.jl (hunk 1 above)
Comment (maleadt) [reply src/communicator.jl:140]: Doesn't seem worth an additional API:

```julia

julia> get_stream(i) = device!(i) do
              stream()
       end

julia> device!(0)
CuDevice(0): NVIDIA H100 PCIe

julia> @benchmark device!(1)
BenchmarkTools.Trial: 10000 samples with 991 evaluations.
 Range (min … max):  41.685 ns … 60.323 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     42.775 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   42.723 ns ±  0.463 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

       ▁▂      ▁     ▅▅▁    █▂                                ▁
  ▅▅▆▇█████▆▆▅██▃▁▄▄▅███▇▆▁▁██▇▄▁▁▁▃▃▁▁▁▁▆▁▁▁▁▁▇▅▁▁▁▁▁▁▁▁▁▁▁▅ █
  41.7 ns      Histogram: log(frequency) by time      44.1 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**Comment by ToucheSir** [comment]
FWIW I've had no issue using libraries which use NCCL on systems with only two devices. Maybe a local environment or config issue?

**Comment by maleadt** [comment]
I'll try setting up a 2 GPU worker.

File: src/communicator.jl (hunk 1 above)
Comment (maleadt) [reply src/communicator.jl:140]: I changed to CuDevice while querying the stream as shown above; feel free to push additional changes.

**Pull request merged**